client/rest-high-level/src/main/java/org/elasticsearch/client/watcher/GetWatchResponse.java

```diff
@@ -0,0 +1,155 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.elasticsearch.client.watcher;
+
+import org.elasticsearch.common.ParseField;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.lucene.uid.Versions;
+import org.elasticsearch.common.xcontent.ConstructingObjectParser;
+import org.elasticsearch.common.xcontent.DeprecationHandler;
+import org.elasticsearch.common.xcontent.NamedXContentRegistry;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.common.xcontent.XContentParser;
+import org.elasticsearch.common.xcontent.XContentType;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.Map;
+import java.util.Objects;
+
+public class GetWatchResponse {
+    private final String id;
+    private final long version;
+    private final WatchStatus status;
+
+    private final BytesReference source;
+    private transient Map<String, Object> sourceAsMap;
+
+    /**
+     * Ctor for missing watch
+     */
+    public GetWatchResponse(String id) {
+        this(id, Versions.NOT_FOUND, null, null);
+    }
+
+    public GetWatchResponse(String id, long version, WatchStatus status, BytesReference source) {
+        this.id = id;
+        this.version = version;
+        this.status = status;
+        this.source = source;
+    }
+
+    public String getId() {
+        return id;
+    }
+
+    public long getVersion() {
+        return version;
+    }
+
+    public boolean isFound() {
+        return version != Versions.NOT_FOUND;
+    }
+
+    public WatchStatus getStatus() {
+        return status;
+    }
+
+    /**
+     * Returns the {@link XContentType} of the source
+     */
+    public XContentType getContentType() {
+        return XContentType.JSON;
+    }
+
+    /**
+     * Returns the serialized watch
+     */
+    public BytesReference getSource() {
+        return source;
+    }
+
+    /**
+     * Returns the source as a map
+     */
+    public Map<String, Object> getSourceAsMap() throws IOException {
+        if (sourceAsMap == null && source != null) {
```

Comment:
A couple questions:
- Sorry if I'm missing something, but why don't we use `XContentHelper#convertToMap` here?
- For me it'd be better to avoid introducing the cached `sourceAsMap` variable, and just do the conversion every time as we do in `GetFieldMappingsResponse` and `IndexRequest`. That way this object can remain immutable, and we don't need to reason about thread safety at all.